theseus/optimizer/autograd/baspacho_sparse_autograd.py

```diff
@@ -0,0 +1,199 @@
+
+    # Let v row vector, and w column vector of dimension n, m, and
+    # A an nxm matrix. Then
+    #   v * A * w = Sum(v_i * A_ij * w_j) = [v (X) w] . A
+    # Where by [v (X) w] we mean the nxm matrix which is the
+    # tensor product of v and w, and "." is the componentwise
+    # dot product of the two nxm matrices.
+    #
+    # Now, we have
+    #      At * A * x = At * b
+    #
+    # Therefore if A, b, x are parametrized (eg. by u) we have deriving
+    #
+    # (i)  At'*A*x + At*A'*x + At*A*x' = At'*b + At*b'
+    #
+    # indicating A'=dA/du, b'=db/du, x'=dx/du
+    #
+    # Now, assume we have a function f of x, and G = df/dx be the
+    # gradient that we consider a row vector, so G*x' is df/du.
+    #
+    # To compute df/db and df/dA, make x' explicit in the (i):
+    #
+    #      x' = (At * A)^{-1} (At*b' + At'*b - At'*A*x - At*A'*x)
+    #
+    # So multiplying by the row vector G we have
+    #
+    #   G*x' = G*(At*A)^{-1}*At * b' + G*(At*A)^{-1} * (At'*b - At'*A*x - At*A'*x)
+    #        = H * At * b' + H * At' * (b - A*x) - H * At * A' * x
+    #        = (H*At) * b' + [H (X) (b-A*x)] . At' - [H*At (X) x] . A'
+    #        = (H*At) * b' + [(b-A*x) (X) H - H*At (X) x] . A'
+    # after putting H = G*(At*A)^{-1} for convenience.
+    #
+    # Therefore after switching to column vectors we have
+    #   df/db = A*H
+    # (where H = (At*A)^{-1}*G), while
+    #   df/dA = (b-A*x) (X) H - A*H (X) x
+    # The two tensor products means that to compute the gradient of
+    # a block of A we have to multiply entries taken from (b-A*x) and H,
+    # and blocks taken from A*H and x.
+    #
+    # Here we assume we are provided x and H after the linear solver
+    # has been applied to Atb and the gradient G.
+
+    # With (large) multiplicative damping, as above with extra terms:
+    #      x' = ... - (AtA_damped)^{-1} * alpha*AtA_diag'*x
+    # So multiplying by the row vector G we have
+    #      G*x' = ... - H * alpha * AtA_diag' * x
+    # Note that '...' the part multiplying H[j]*alpha*x[j] is AtA_diag'[j], ie
+    # 2 times the scalar product of A's an (A')'s j-th colum. Therefore
+    # (A')'s j-th colum is multiplying A's j-th colum by 2*H[j]*alpha*x[j]
+    @staticmethod
+    def backward(ctx, grad_output):
+        batch_size = grad_output.shape[0]
+        targs = {"dtype": grad_output.dtype, "device": grad_output.device}
+
+        H = grad_output.clone()
+        ctx.numeric_decomposition.solve(H)  # solve in place
+
+        A_args = ctx.sparse_structure.num_cols, ctx.A_rowPtr, ctx.A_colInd, ctx.A_val
+        AH = mat_vec(batch_size, *A_args, H)
+        b_Ax = ctx.b - mat_vec(batch_size, *A_args, ctx.x)
+
+        # now we fill values of a matrix with structure identical to A with
+        # selected entries from the difference of tensor products:
+        #   b_Ax (X) H - AH (X) x
+        # NOTE: this row-wise manipulation can be much faster in C++ or Cython
+        A_colInd = ctx.sparse_structure.col_ind
```

Comment:
From a quick look the code below this line seems to be equal to the one in `lu_cuda_sparse_autograd.py`. If so, can we encapsulate this into a function that both of these call?

Reply:
I think we may have an issue for this already?

Reply:
Yes, definitely to be addressed. only objection is that the code has several similar parts but also subtle differences, so unifying and making it generic might end up being harder said than done. (For instance, data saved in the ctx are quite different for all the autograd classes.) If ok I propose solving this in a future PR, as it might require iterating several times before I get it right!

Reply:
Ok to resolve? Then getting batch to this in a future PR. Doesn't look like the best thing to do delaying this PR to solve an issue already present in the code...

Reply:
Yes, OK to resolve.